ingest/downloader/downloader.py

```diff
@@ -0,0 +1,98 @@
+                worksheet_name = self.get_concrete_entity(content)
+                row = {f'{worksheet_name}.uuid': entity['uuid']['uuid']}
+
+            if not worksheet_name:
+                raise Exception('There should be a worksheet name')
+
+            user_friendly_worksheet_name = self._format_worksheet_name(worksheet_name)
+
+            rows = self.workbook.get(user_friendly_worksheet_name, [])
+            self.workbook[user_friendly_worksheet_name] = rows
+            self._flatten_object(content, row, parent_key=worksheet_name)
+            rows.append(row)
+
+    def _flatten_object(self, object: dict, flattened_object: dict, parent_key: str = ''):
+        if isinstance(object, dict):
+            for key in object:
+                full_key = f'{parent_key}.{key}' if parent_key else key
+                if key in EXCLUDE_KEYS:
+                    continue
+                value = object[key]
+                if isinstance(value, dict) or isinstance(value, list):
+                    self._flatten_object(value, flattened_object, parent_key=full_key)
+                else:
+                    flattened_object[full_key] = str(value)
+        elif isinstance(object, list):
+            if self.is_object_list(object):
+                self._flatten_object_list(object, parent_key)
+            else:
+                stringified = [str(e) for e in object]
+                flattened_object[parent_key] = '||'.join(stringified)
+
+    def _format_worksheet_name(self, worksheet_name):
+        names = worksheet_name.split('.')
+        names = [n.replace('_', ' ') for n in names]
+        new_worksheet_name = ' - '.join([n.capitalize() for n in names])
+        return new_worksheet_name
+
+    def is_object_list(self, content):
+        return content and isinstance(content[0], dict)
+
+    @staticmethod
+    def get_concrete_entity(content):
+        return content.get('describedBy').rsplit('/', 1)[-1]
+
+    def create_workbook(self, input_json: dict) -> Workbook:
+        workbook = Workbook()
+        workbook.remove(workbook.active)
+
+        for ws_title, ws_elements in input_json.items():
+            worksheet: Worksheet = workbook.create_sheet(title=ws_title)
+            self.add_worksheet_content(worksheet, ws_elements)
+
+        return workbook
+
+    def add_worksheet_content(self, worksheet, ws_elements: dict):
+        is_header = True
+        if isinstance(ws_elements, list):
+            for content in ws_elements:
+                self.add_row_content(worksheet, content, is_header)
+                is_header = False
+                self.row += 1
+        else:
+            self.add_row_content(worksheet, ws_elements)
+
+    def add_row_content(self, worksheet, content, is_header=True):
+        col = 1
+        for header, cell_value in content.items():
+            if is_header:
+                self.row = 1
+                worksheet.cell(row=self.row, column=col, value=header)
+                self.row += 1
```

Comment:
Are you sure about this indentation?

Reply:
yes and there are working tests for it 😏